### PR TITLE
[NTUSER][USER32] Implement IME status

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

Implementing Japanese input...
JIRA issue: [CORE-11700](https://jira.reactos.org/browse/CORE-11700)

## Proposed changes

- Add `IntCheckImeShowStatus`, `IntSendMessageToUI`, `IntSendOpenStatusNotify`, `IntNotifyImeShowStatus`, and `xxxBroadcastImeShowStatusChange` helper functions.
- Renaming: `s/X_ROUTINE_IMESHOWSTATUSCHANGE/TWOPARAM_ROUTINE_IMESHOWSTATUSCHANGE/`
- Implement `NtUserCallNoParam.NOPARAM_ROUTINE_GETIMESHOWSTATUS`.
- Implement `NtUserCallHwndParamLock.TWOPARAM_ROUTINE_IMESHOWSTATUSCHANGE`.
- Fix hung in `User32GetTopLevelWindow` and rename it as `IntGetTopLevelWindow`.

## TODO

- [x] Build.
- [x] Do tests.

### CODE REVIEW

**Comment by katahiromz** [comment]
Branch: refs/pull/4472/head
https://build.reactos.org/
https://reactos.org/testman/

BEFORE: de31647
AFTER: dfaedc5

Test KVM BEFORE: de31647
Test KVM AFTER: dfaedc5
https://reactos.org/testman/compare.php?ids=???,???

Test VBox BEFORE: de31647
Test VBox AFTER: dfaedc5
https://reactos.org/testman/compare.php?ids=???,???

File: win32ss/user/ntuser/ime.c [hunk 1]
```diff
@@ -2101,4 +2102,289 @@ BOOL FASTCALL IntImeCanDestroyDefIME(PWND pImeWnd, PWND pwndTarget)
...
+    if (pti != ptiIME)
+        KeAttachProcess(&(ptiIME->ppi->peProcess->Pcb));
+
+    // Get an IMEUI and check whether hwndIMC is valid and update fShowStatus
+    _SEH2_TRY
+    {
+        ProbeForWrite(pImeWnd, sizeof(IMEWND), 1);
+        SafeImeUI = *((PIMEWND)pImeWnd)->pimeui;
+
+        bShow = gfIMEShowStatus && SafeImeUI.fCtrlShowStatus;
```
Comment (learn-more) [inline win32ss/user/ntuser/ime.c:2325]: Can this be called when `gfIMEShowStatus` is still `-1`

Comment (katahiromz) [reply win32ss/user/ntuser/ime.c]: Fixed.

**Comment by HBelusca** [comment]
Does this also fix the handles leak that make the programs to not completely quit?

**Comment by katahiromz** [comment]
>Does this also fix the handles leak

I think not so.

**Comment by jimtabor** [comment]
Please review https://reactos.org/wiki/Techwiki:Win32k/apfnSimpleCall

See NOPARAM_ROUTINE_GETIMESHOWSTATUS

Why is there a X_ROUTINE_Zyx ?

**Comment by HBelusca** [comment]
> > Does this also fix the handles leak
> 
> I think not so. Can you report it to Jira?

Well, it affects the whole testing infrastructure (and any manual usages we do), so I would expect that you should be already actively trying to fix that in order not to delay things...

File: win32ss/user/ntuser/ime.c [hunk 2]
```diff
@@ -2101,4 +2102,289 @@ BOOL FASTCALL IntImeCanDestroyDefIME(PWND pImeWnd, PWND pwndTarget)
...
+BOOL FASTCALL IntSetIMEShowStatus(BOOL bShow)
+{
+    PWND pwndFocus, pImeWnd;
+    PTHREADINFO pti;
+    USER_REFERENCE_ENTRY Ref;
+
+    if (gfIMEShowStatus == bShow)
+        return TRUE;
+
+    gfIMEShowStatus = bShow;
```
Comment (HBelusca) [inline win32ss/user/ntuser/ime.c:2366]: Should this be always done? Or only if IMM mode is enabled? (see the check after)

File: win32ss/user/ntuser/ime.c [hunk 3]
```diff
@@ -2101,4 +2102,289 @@ BOOL FASTCALL IntImeCanDestroyDefIME(PWND pImeWnd, PWND pwndTarget)
...
+    BOOL bShow, bSendNotify = FALSE;
+    IMEUI SafeImeUI;
+
+    if (!IS_IMM_MODE() || (pImeWnd->state2 & WNDS2_INDESTROY))
+        return;
+
+    pti = PsGetCurrentThreadWin32Thread();
+    ptiIME = pImeWnd->head.pti;
+
+    // Attach the process if necessary
```
Comment (HBelusca) [inline win32ss/user/ntuser/ime.c:2315]:
```suggestion
    // Attach to the process if necessary
```

File: win32ss/user/ntuser/ime.c [hunk 4]
```diff
@@ -2101,4 +2102,289 @@ BOOL FASTCALL IntImeCanDestroyDefIME(PWND pImeWnd, PWND pwndTarget)
...
+    }
+    _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+    {
+        if (pti != ptiIME)
+            KeDetachProcess();
+        return;
+    }
+    _SEH2_END;
+
+    // Detach the process if necessary
```
Comment (HBelusca) [inline win32ss/user/ntuser/ime.c:2345]:
```suggestion
    // Detach from the process if necessary
```

File: win32ss/user/ntuser/ime.c [hunk 5]
```diff
@@ -2101,4 +2102,289 @@ BOOL FASTCALL IntImeCanDestroyDefIME(PWND pImeWnd, PWND pwndTarget)
...
+// Send the open status notification.
+// Win: xxxSendOpenStatusNotify
+VOID FASTCALL
+IntSendOpenStatusNotify(PTHREADINFO ptiIME, PIMEUI pimeui, PWND pWnd, BOOL bOpen)
+{
+    WPARAM wParam = (bOpen ? IMN_OPENSTATUSWINDOW : IMN_CLOSESTATUSWINDOW);
+    PTHREADINFO ptiWnd = pWnd->head.pti;
+    USER_REFERENCE_ENTRY Ref;
+
+    if (ptiWnd->dwExpWinVer >= 0x400 && pWnd->hImc)
```
Comment (HBelusca) [inline win32ss/user/ntuser/ime.c:2287]:
```suggestion
    if (ptiWnd->dwExpWinVer >= WINVER_WINNT4 && pWnd->hImc)
```

File: win32ss/user/ntuser/ime.c [hunk 6]
```diff
@@ -2101,4 +2102,289 @@ BOOL FASTCALL IntImeCanDestroyDefIME(PWND pImeWnd, PWND pwndTarget)
...
+        InterlockedDecrement(&pimeui->nCntInIMEProc);
+    }
+    _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+    {
+        goto Quit;
+    }
+    _SEH2_END;
+
+Quit:
+    // Detach the process if necessary
```
Comment (HBelusca) [inline win32ss/user/ntuser/ime.c:2271]:
```suggestion
    // Detach from the process if necessary
```

File: win32ss/user/ntuser/ime.c [hunk 7]
```diff
@@ -2101,4 +2102,289 @@ BOOL FASTCALL IntImeCanDestroyDefIME(PWND pImeWnd, PWND pwndTarget)
...
+
+    // Detach the process if necessary
+    if (bDifferent)
+        KeDetachProcess();
+
+    UserRefObjectCo(pwndUI, &Ref);
+    ret = co_IntSendMessage(UserHMGetHandle(pwndUI), uMsg, wParam, lParam);
+    UserDerefObjectCo(pwndUI);
+
+    // Attach the process if necessary
```
Comment (HBelusca) [inline win32ss/user/ntuser/ime.c:2254]:
```suggestion
    // Attach to the process if necessary
```

File: win32ss/user/ntuser/ime.c [hunk 8]
```diff
@@ -2101,4 +2102,289 @@ BOOL FASTCALL IntImeCanDestroyDefIME(PWND pImeWnd, PWND pwndTarget)
...
+        ProbeForWrite(&pimeui->nCntInIMEProc, sizeof(LONG), 1);
+        InterlockedIncrement(&pimeui->nCntInIMEProc);
+    }
+    _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+    {
+        goto Quit;
+    }
+    _SEH2_END;
+
+    // Detach the process if necessary
```
Comment (HBelusca) [inline win32ss/user/ntuser/ime.c:2246]:
```suggestion
    // Detach from the process if necessary
```

File: win32ss/user/ntuser/ime.c [hunk 9]
```diff
@@ -2101,4 +2102,289 @@ BOOL FASTCALL IntImeCanDestroyDefIME(PWND pImeWnd, PWND pwndTarget)
...
+LRESULT FASTCALL
+IntSendMessageToUI(PTHREADINFO ptiIME, PIMEUI pimeui, UINT uMsg, WPARAM wParam, LPARAM lParam)
+{
+    PWND pwndUI;
+    LRESULT ret = 0;
+    IMEUI SafeImeUI;
+    BOOL bDifferent = FALSE;
+    USER_REFERENCE_ENTRY Ref;
+
+    // Attach the process if necessary
```
Comment (HBelusca) [inline win32ss/user/ntuser/ime.c:2211]:
```suggestion
    // Attach to the process if necessary
```

File: win32ss/user/ntuser/ime.c [hunk 10]
```diff
@@ -2101,4 +2102,289 @@ BOOL FASTCALL IntImeCanDestroyDefIME(PWND pImeWnd, PWND pwndTarget)
...
+    _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+    {
+        pwndUI = NULL;
+    }
+    _SEH2_END;
+
+    if (!pwndUI)
+        goto Quit;
+
+    // Lock the IME processor
```
Comment (HBelusca) [inline win32ss/user/ntuser/ime.c:2234]: Looks more like a reference count is taken to say the code execution is in an IME procedure....

File: win32ss/user/ntuser/ime.c [hunk 11]
```diff
@@ -2101,4 +2102,289 @@ BOOL FASTCALL IntImeCanDestroyDefIME(PWND pImeWnd, PWND pwndTarget)
...
+                pwndIMC = NULL;
+            }
+        }
+        _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+        {
+            pwndIMC = NULL;
+        }
+        _SEH2_END;
+
+        // Detach the process if necessary
```
Comment (HBelusca) [inline win32ss/user/ntuser/ime.c:2176]:
```suggestion
        // Detach from the process if necessary
```

File: win32ss/user/ntuser/ime.c [hunk 12]
```diff
@@ -2101,4 +2102,289 @@ BOOL FASTCALL IntImeCanDestroyDefIME(PWND pImeWnd, PWND pwndTarget)
...
+        }
+
+        pimeui = ((PIMEWND)pwndNode)->pimeui;
+        if (!pimeui || pimeui == (PIMEUI)-1)
+            continue;
+
+        if (pti && pti != pwndNode->head.pti)
+            continue;
+
+        // Attach the process if necessary
```
Comment (HBelusca) [inline win32ss/user/ntuser/ime.c:2146]:
```suggestion
        // Attach to the process if necessary
```

File: win32ss/user/ntuser/ime.c [hunk 13]
```diff
@@ -2101,4 +2102,289 @@ BOOL FASTCALL IntImeCanDestroyDefIME(PWND pImeWnd, PWND pwndTarget)
...
+            continue;
+
+        if (pwndNode->pcls->atomClassName != gpsi->atomSysClass[ICLS_IME] ||
+            (pwndNode->state2 & WNDS2_INDESTROY))
+        {
+            continue;
+        }
+
+        pimeui = ((PIMEWND)pwndNode)->pimeui;
+        if (!pimeui || pimeui == (PIMEUI)-1)
```
Comment (HBelusca) [inline win32ss/user/ntuser/ime.c:2140]: For what purpose could the `pimeui` member be assigned the value -1 (somewhere else in the code)?

File: win32ss/user/ntuser/ime.c (hunk 10 above)
Comment (katahiromz) [reply win32ss/user/ntuser/ime.c:2234]: See also `ImeWndProc_common` at File `win32ss/user/user32/misc/imm.c`.

Comment (HBelusca) [reply win32ss/user/ntuser/ime.c:2234]: Ah, ok, that's a *recursion count*.

File: win32ss/user/ntuser/ime.c (hunk 13 above)
Comment (katahiromz) [reply win32ss/user/ntuser/ime.c:2140]: I don't know well. Something validation?

File: win32ss/user/ntuser/ime.c (hunk 2 above)
Comment (katahiromz) [reply win32ss/user/ntuser/ime.c:2366]: Fixed.

File: win32ss/user/ntuser/ime.c (hunk 10 above)
Comment (katahiromz) [reply win32ss/user/ntuser/ime.c]: Fixed these comments.

File: win32ss/user/ntuser/ime.c (hunk 13 above)
Comment (HBelusca) [reply win32ss/user/ntuser/ime.c:2140]: I don't know. You write that code, so you should know better than me what you are writing and for which purpose 😂😂

**Comment by katahiromz** [comment]
@jimtabor 
> Why is there a X_ROUTINE_Zyx ?

There is `xxxBroadcastImeShowStatusChange` for it.

**Comment by katahiromz** [comment]
Branch: refs/pull/4472/head
https://build.reactos.org/
https://reactos.org/testman/

BEFORE: 4a748a3 https://build.reactos.org/#/builders/9/builds/32068
AFTER: 4b65740 https://build.reactos.org/#/builders/9/builds/32069

Test KVM BEFORE: 4a748a3 id=???
Test KVM AFTER: 4b65740 id=???
https://reactos.org/testman/compare.php?ids=???,???

Test VBox BEFORE: 4a748a3 id=???
Test VBox AFTER: 4b65740 id=???
https://reactos.org/testman/compare.php?ids=???,???

**Comment by katahiromz** [comment]
Ready Go!

Branch: refs/pull/4472/head
https://build.reactos.org/
https://reactos.org/testman/

BEFORE: a849125
AFTER: 2bb40de

Test KVM BEFORE: a849125 id=???
Test KVM AFTER: 2bb40de id=???
https://reactos.org/testman/compare.php?ids=???,???

Test VBox BEFORE: a849125 id=???
Test VBox AFTER: 2bb40de id=???
https://reactos.org/testman/compare.php?ids=???,???

**Pull request merged**